webhook/index.js

```diff
@@ -0,0 +1,71 @@
+var bunyan = require('bunyan');
+var request = require('superagent');
+var noCache = require('superagent-no-cache')
+
+var log = bunyan.createLogger({name: "webhook"});
+
+var RETRY_DELAY = 60;
+
+function assertMessageIsWebhookEvent(msg) {
+    return !!msg.event && !!msg.url && !!msg.createdAt;
+}
+
+require('amqplib/callback_api').connect(
+    'amqp://localhost',
+    (err, conn) => {
+        if (err != null) {
+            return log.error({error: err}, 'queue connection error');
+        }
+
+        conn.createChannel((err, ch) => {
+            if (err != null) {
+                return log.error({error: err}, 'queue channel creation error');
+            }
+
+            ch.assertQueue('webhooks');
+            ch.consume('webhooks', (msg) => {
+                if (msg !== null) {
+                    msg = JSON.parse(msg.content.toString());
+
+                    if (!assertMessageIsWebhookEvent(msg)) {
+                        // FIXME: major error here, should we stop the service ?
+                        log.error(
+                            { message: msg },
+                            'assertion failed: queue message is not a valid webhook'
+                        );
+                        return ;
+                    }
+
+                    log.info({ webhook: msg }, 'handling webhook');
+
+                    if (!!msg.lastTriedAt && (msg.lastTriedAt - msg.lastTriedAt > RETRY_DELAY)) {
+                        log.info({ webhook: msg }, 'delaying webhook');
+                        ch.nack(msg, false, true);
```

Comment:
The problem with this approach is we will loop over the same messages so long as the target server is unavailable.

A solution could be to use a second queue to sent unprocessed messages to. After a given TTL, the messages would then be sent back to the main queue for retry. This is called a "dead-letter exchange" in RabbitMQ lingua. See an example here: http://yuserinterface.com/dev/2013/01/08/how-to-schedule-delay-messages-with-rabbitmq-using-a-dead-letter-exchange/

Additionally, when a message reaches a second, longer TTL (like a day or two), it should be rejected so as to not clutter the queue forever.